## <b><font color='darkblue'>Preface</font></b>
([source](https://nanonets.com/blog/langchain/)) <font size='3ptx'>At its core, [**LangChain**](https://www.langchain.com/) is an innovative framework tailored for crafting applications that leverage the capabilities of language models</font>. It's a toolkit designed for developers to create applications that are context-aware and capable of sophisticated reasoning. <b>This framework is highly relevant when discussing [Retrieval-Augmented Generation](https://nanonets.com/blog/what-is-retrieval-augmented-generation-rag/) (RAG), a concept that enhances the effectiveness of language models in generating responses based on retrieved data</b>.

<b>This means LangChain applications can understand the context, such as prompt instructions or content grounding responses and use [**Large Language Models**](https://nanonets.com/blog/what-are-large-language-models/) for complex reasoning tasks, like deciding how to respond or what actions to take</b>. LangChain represents a unified approach to developing intelligent applications, simplifying the journey from concept to execution with its diverse components.

In [5]:
from IPython.display import display
from IPython.display import Markdown
import textwrap

### <b><font color='darkgreen'>Understanding LangChain</font></b>
[**LangChain**](https://python.langchain.com/docs/tutorials/) is much more than just a framework; it's a full-fledged ecosystem comprising several integral parts ([more](https://python.langchain.com/v0.2/docs/concepts/)): 
* **Firstly, there are the <font color='darkblue'>LangChain Libraries</font>, available in both Python and JavaScript**. These libraries are the backbone of LangChain, offering interfaces and integrations for various components. They provide a basic runtime for combining these components into cohesive chains and agents, along with ready-made implementations for immediate use.
* **Next, we have <font color='darkblue'>[LangChain Templates](https://templates.langchain.com/)</font>**. These are a collection of deployable reference architectures tailored for a wide array of tasks. Whether you're building a chatbot or a complex analytical tool, these templates offer a solid starting point. If you're looking for insights into how LangChain can integrate with other systems, including leveraging powerful tools like LLamaIndex, this blog will be of great help.
* **<font color='darkblue'>LangServe</font> steps in as a versatile library for deploying LangChain chains as REST APIs**. This tool is essential for turning your LangChain projects into accessible and scalable web services.
* **Lastly, <font color='darkblue'>LangSmith</font> serves as a developer platform**. It's designed to debug, test, evaluate, and monitor chains built on any LLM framework. The seamless integration with LangChain makes it an indispensable tool for developers aiming to refine and perfect their applications. For more information on how you can streamline and automate your workflows with LLMs, check out our post on Leveraging LLMs to Streamline and Automate Your Workflows.

![components](https://python.langchain.com/v0.2/svg/langchain_stack_062024.svg)

<b>Together, these components empower you to develop, productionize, and deploy applications with ease</b>. With LangChain, you start by writing your applications using the libraries, referencing templates for guidance. LangSmith then helps you in inspecting, testing, and monitoring your chains, ensuring that your applications are constantly improving and ready for deployment. Finally, with LangServe, you can easily transform any chain into an API, making deployment a breeze.

<b>In the next sections, we will delve deeper into how to set up LangChain and begin your journey in creating intelligent, language model-powered applications.</b>

### <b><font color='darkgreen'>Installation and Setup</font></b>
<b>Are you ready to dive into the world of LangChain? Setting it up is straightforward, and this guide will walk you through the process step-by-step.</b>

The first step in your LangChain journey is to install it. You can do this easily using pip or conda. Run the following command in your terminal:

In [2]:
# pip install langchain
!pip freeze | grep 'langchain'

langchain==0.3.17
langchain-anthropic==0.2.1
langchain-cli==0.0.35
langchain-community==0.3.16
langchain-core==0.3.33
langchain-experimental==0.0.62
langchain-google-genai==2.0.9
langchain-google-vertexai==1.0.9
langchain-groq==0.1.3
langchain-openai==0.3.3
langchain-text-splitters==0.3.5
langchainhub==0.1.21
openinference-instrumentation-langchain==0.1.27


LangChain CLI is a handy tool for working with LangChain templates and LangServe projects. To install the LangChain CLI, use:

In [3]:
# pip install langchain-cli
!pip freeze | grep 'langchain-cli'

langchain-cli==0.0.35


LangServe is essential for deploying your LangChain chains as a REST API. It gets installed alongside the LangChain CLI.

LangChain often requires integrations with model providers, data stores, APIs, etc. For this example, we'll use OpenAI's model APIs. Install the OpenAI Python package using:

In [4]:
# pip install -U openai
!pip freeze | grep 'openai'

langchain-openai==0.3.3
llama-index-agent-openai==0.4.0
llama-index-embeddings-openai==0.3.1
llama-index-llms-openai==0.3.2
llama-index-multi-modal-llms-openai==0.3.0
llama-index-program-openai==0.3.1
llama-index-question-gen-openai==0.3.0
openai==1.60.2


In [5]:
# pip install -q --upgrade google-generativeai langchain-google-genai python-dotenv
!pip freeze | grep 'langchain-google-genai'

langchain-google-genai==2.0.9


To access the API, set your OpenAI API key as an environment variable:
```shell
$ export OPENAI_API_KEY="your_api_key"
```

In [6]:
from dotenv import load_dotenv, find_dotenv
import os


os.environ['ALLOW_RESET'] = 'TRUE'
_ = load_dotenv(find_dotenv(os.path.expanduser('~/.env')))
assert os.environ.get('OPENAI_API_KEY', '')

<b>LangChain allows for the creation of language model applications through modules</b>. These modules can either stand alone or be composed for complex use cases. These modules are:
* **Model I/O**: Facilitates interaction with various language models, handling their inputs and outputs efficiently.
* **Retrieval**: Enables access to and interaction with application-specific data, crucial for dynamic data utilization.
* **Agents**: Empower applications to select appropriate tools based on high-level directives, enhancing decision-making capabilities.
* **Chains**: Offers pre-defined, reusable compositions that serve as building blocks for application development.
* **Memory**: Maintains application state across multiple chain executions, essential for context-aware interactions.

Each module targets specific development needs, making LangChain a comprehensive toolkit for creating advanced language model applications.

Along with the above components, we also have <b><font color='darkblue'>LangChain Expression Language</font> (LCEL), which is a declarative way to easily compose modules together, and this enables the chaining of components using a universal Runnable interface</b>.

LCEL looks something like this:
```python
from langchain.chat_models import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain.schema import BaseOutputParser

# Example chain
chain = ChatPromptTemplate() | ChatOpenAI() | CustomOutputParser()
```

Now that we have covered the basics, we will continue on to:
* Dig deeper into each Langchain module in detail.
* Learn how to use LangChain Expression Language.
* Explore common LangChain examples and implement them.
* Explore common use cases
* Deploy an end-to-end application with LangServe.
* Check out LangSmith for debugging, testing, and monitoring.

<a id='agenda'></a>
### <b><font color='darkgreen'>Agenda</font></b>
* <font size='3ptx'>[**Module I : Model I/O**](#sect_1)</font>
* <font size='3ptx'>[**Module II : Retrieval**](#sect_2)</font>
* <font size='3ptx'>[**Module III : Agents**](#sect_3)</font>
* <font size='3ptx'>[**Module IV : Chains**](#sect_4)</font>
* <font size='3ptx'>[**Module V : Memory**](#sect_5)</font>

<a id='sect_1'></a>
## <b><font color='darkblue'>Module I : Model I/O</font></b>
<b><font size='3ptx'>In LangChain, the core element of any application revolves around the language model</font>. This module provides the essential building blocks to interface effectively with any language model, ensuring seamless integration and communication.</b>

Key Components of Model I/O
1. <b>LLMs and Chat Models</b> (used interchangeably)
   * LLMs:
      - Definition: Pure text completion models.
      - Input/Output: Take a text string as input and return a text string as output.
   * Chat Models
      - Definition: Models that use a language model as a base but differ in input and output formats.
      - Input/Output: Accept a list of chat messages as input and return a Chat Message.
2. **Prompts**: Templatize, dynamically select, and manage model inputs. Allows for the creation of flexible and context-specific prompts that guide the language model's responses.
3. **Output Parsers**: Extract and format information from model outputs. Useful for converting the raw output of language models into structured data or specific formats needed by the application.

### <b><font color='darkgreen'>[LLMs](https://nanonets.com/blog/what-are-large-language-models/)</font></b>
<b><font size='3ptx'>LangChain's integration with Large Language Models (LLMs) like OpenAI, Cohere, and Hugging Face is a fundamental aspect of its functionality</font>. LangChain itself does not host LLMs but offers a uniform interface to interact with various LLMs.</b>

This section provides an overview of using the OpenAI LLM wrapper in LangChain ([details](https://python.langchain.com/docs/integrations/chat/google_generative_ai/)), applicable to other LLM types as well. We have already installed this in the "Getting Started" section. Let us initialize the LLM.

In [7]:
from langchain_openai import OpenAI

llm = OpenAI()

LLMs implement the [**Runnable interface**](https://python.langchain.com/docs/expression_language/interface), the basic building block of the LangChain Expression Language (LCEL). This means they support:
* [**`Invoked`**](https://python.langchain.com/docs/how_to/lcel_cheatsheet/#invoke-a-runnable): A single input is transformed into an output.
* [**`Batched`**](https://python.langchain.com/docs/how_to/lcel_cheatsheet/#batch-a-runnable): Multiple inputs are efficiently transformed into outputs.
* [**`Streamed`**](https://python.langchain.com/docs/how_to/lcel_cheatsheet/#stream-a-runnable): Outputs are streamed as they are produced.
* **`Inspected`**: Schematic information about Runnable's input, output, and configuration can be accessed.
* **`Composed`**: Multiple Runnables can be composed to work together using [**the LangChain Expression Language**](https://python.langchain.com/docs/concepts/lcel/) (LCEL) to create complex pipelines.

LLMs accept **strings** as inputs, or objects which can be coerced to string prompts, including
* **List[BaseMessage]**
* **PromptValue**
* **More**

Let us look at some examples

In [8]:
response = llm.invoke("List the seven wonders of the world.")

In [9]:
Markdown(response)



1. The Great Wall of China
2. The Taj Mahal
3. The Colosseum
4. The Christ the Redeemer statue
5. Machu Picchu
6. The Great Pyramid of Giza
7. Chichen Itza

You can alternatively call the stream method to stream the text response.

In [10]:
response_chunks = []
for chunk in llm.stream("Where were the 2012 Olympics held?"):
    print(chunk, end="", flush=True)
    response_chunks.append(chunk)



The 2012 Olympics were held in London, England.

In [11]:
Markdown(''.join(response_chunks))



The 2012 Olympics were held in London, England.

### <b><font color='darkgreen'>Chat Models</font></b>
<b><font size='3ptx'>LangChain's integration with [chat models](https://python.langchain.com/docs/integrations/chat/), a specialized variation of language models, is essential for creating interactive chat applications</font>. While they utilize language models internally, chat models present a distinct interface centered around chat messages as inputs and outputs</b>. 

This section provides a detailed overview of using Gemini's chat model ([**ChatGoogleGenerativeAI**](https://python.langchain.com/docs/integrations/chat/google_generative_ai/)) in LangChain.

In [12]:
import getpass
import os

if "GOOGLE_API_KEY" not in os.environ:
    os.environ["GOOGLE_API_KEY"] = getpass.getpass("Enter your Google AI API key: ")

In [13]:
from langchain_google_genai import ChatGoogleGenerativeAI

chat = ChatGoogleGenerativeAI(
    model="gemini-1.5-pro",
    temperature=0,
    max_tokens=None,
    timeout=None,
    max_retries=2,
    # other params...
)

Chat models in LangChain work with different message types such as:
* [**AIMessage**](https://python.langchain.com/api_reference/core/messages/langchain_core.messages.ai.AIMessage.html): AIMessage is returned from a chat model as a response to a prompt.
* [**HumanMessage**](https://python.langchain.com/api_reference/core/messages/langchain_core.messages.human.HumanMessage.html): HumanMessages are messages that are passed in from a human to the model
* [**SystemMessage**](https://python.langchain.com/api_reference/core/messages/langchain_core.messages.system.SystemMessage.html): Message for priming AI behavior.
* [**FunctionMessage**](https://python.langchain.com/api_reference/core/messages/langchain_core.messages.function.FunctionMessage.html): Message for passing the result of executing a tool back to a model.
* [**ChatMessage**](https://python.langchain.com/api_reference/core/messages/langchain_core.messages.chat.ChatMessage.html): Message that can be assigned an arbitrary speaker (i.e. role).
* [**More**](https://python.langchain.com/api_reference/core/messages.html)

In [14]:
from langchain.schema.messages import HumanMessage, SystemMessage

messages = [
    SystemMessage(content="You are Micheal Jordan."),
    HumanMessage(content="Which shoe manufacturer are you associated with?"),
]
response = chat.invoke(messages)

In [15]:
print(response.content)

Nike.  And I'm incredibly proud of the Jordan Brand and its impact on the game and culture.  It's more than just shoes; it's a legacy.


### <b><font color='darkgreen'>Prompts</font></b>
<b><font size='3ptx'>Prompts are essential in guiding language models to generate relevant and coherent outputs</font>. They can range from simple instructions to complex few-shot examples. In LangChain, handling prompts can be a very streamlined process, thanks to several dedicated classes and functions.</b>

LangChain's [**PromptTemplate**](https://python.langchain.com/api_reference/core/prompts/langchain_core.prompts.prompt.PromptTemplate.html) class is a versatile tool for creating string prompts. It uses Python's `str.format` syntax, allowing for dynamic prompt generation. You can define a template with placeholders and fill them with specific values as needed.

In [16]:
from langchain.prompts import PromptTemplate

# Simple prompt with placeholders
prompt_template = PromptTemplate.from_template(
    "Tell me a {adjective} joke about {content}."
)

In [17]:
# Filling placeholders to create a prompt
filled_prompt = prompt_template.format(adjective="funny", content="robots")
print(filled_prompt)

Tell me a funny joke about robots.


For chat models, prompts are more structured, involving messages with specific roles. LangChain offers [**ChatPromptTemplate**](https://python.langchain.com/api_reference/core/prompts/langchain_core.prompts.chat.ChatPromptTemplate.html) for this purpose.

In [18]:
from langchain.prompts import ChatPromptTemplate

# Defining a chat prompt with various roles
chat_template = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a helpful AI bot. Your name is {name}."),
        ("human", "Hello, how are you doing?"),
        ("ai", "I'm doing well, thanks!"),
        ("human", "{user_input}"),
    ]
)

In [19]:
# Formatting the chat prompt
formatted_messages = chat_template.format_messages(name="Bob", user_input="What is your name?")
for message in formatted_messages:
    print(message)

content='You are a helpful AI bot. Your name is Bob.' additional_kwargs={} response_metadata={}
content='Hello, how are you doing?' additional_kwargs={} response_metadata={}
content="I'm doing well, thanks!" additional_kwargs={} response_metadata={}
content='What is your name?' additional_kwargs={} response_metadata={}


This approach allows for the creation of interactive, engaging chatbots with dynamic responses. Both [**PromptTemplate**](https://python.langchain.com/api_reference/core/prompts/langchain_core.prompts.prompt.PromptTemplate.html) and [**ChatPromptTemplate**](https://python.langchain.com/api_reference/core/prompts/langchain_core.prompts.chat.ChatPromptTemplate.html) integrate seamlessly with the LangChain Expression Language (LCEL), enabling them to be part of larger, complex workflows. We will discuss more on this later.

Custom prompt templates are sometimes essential for tasks requiring unique formatting or specific instructions. Creating a custom prompt template involves defining input variables and a custom formatting method. This flexibility allows LangChain to cater to a wide array of application-specific requirements. 

LangChain also supports few-shot prompting, enabling the model to learn from examples. This feature is vital for tasks requiring contextual understanding or specific patterns. **Few-shot prompt templates can be built from a set of examples or by utilizing an Example Selector object. Read more [here](https://python.langchain.com/docs/how_to/few_shot_examples/)**.

### <b><font color='darkgreen'>Output Parsers</font></b>
<b><font size='3ptx'>Output parsers play a crucial role in Langchain, enabling users to structure the responses generated by language models</font>. In this section, we will explore the concept of output parsers and provide code examples using Langchain's [PydanticOutputParser](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.pydantic.PydanticOutputParser.html), [SimpleJsonOutputParser](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.json.SimpleJsonOutputParser.html), [CommaSeparatedListOutputParser](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.list.CommaSeparatedListOutputParser.html), DatetimeOutputParser, and [XMLOutputParser](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.xml.XMLOutputParser.html).</b>

#### <b><font size='3ptx'>PydanticOutputParser</font></b>
Langchain provides the [**PydanticOutputParser**](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.pydantic.PydanticOutputParser.html) for parsing responses into Pydantic data structures. Below is a step-by-step example of how to use it:

In [27]:
from typing import List
from langchain.llms import OpenAI
from langchain.output_parsers import PydanticOutputParser
from langchain.prompts import PromptTemplate
from pydantic import BaseModel, Field, field_validator


# Initialize the language model
model = OpenAI(model_name="gpt-3.5-turbo-instruct", temperature=0.0)

In [28]:
# Define your desired data structure using Pydantic
class Joke(BaseModel):
    setup: str = Field(description="question to set up a joke")
    punchline: str = Field(description="answer to resolve the joke")

    @field_validator("setup")
    def question_ends_with_question_mark(cls, field_value):  # Renamed 'field' to 'field_value' for clarity
        if field_value and field_value[-1] != "?": # Added check for empty string
            raise ValueError("Badly formed question!")
        return field_value

In [29]:
# Set up a PydanticOutputParser
parser = PydanticOutputParser(pydantic_object=Joke)

In [30]:
print(parser.get_format_instructions())

The output should be formatted as a JSON instance that conforms to the JSON schema below.

As an example, for the schema {"properties": {"foo": {"title": "Foo", "description": "a list of strings", "type": "array", "items": {"type": "string"}}}, "required": ["foo"]}
the object {"foo": ["bar", "baz"]} is a well-formatted instance of the schema. The object {"properties": {"foo": ["bar", "baz"]}} is not well-formatted.

Here is the output schema:
```
{"properties": {"setup": {"description": "question to set up a joke", "title": "Setup", "type": "string"}, "punchline": {"description": "answer to resolve the joke", "title": "Punchline", "type": "string"}}, "required": ["setup", "punchline"]}
```


In [31]:
# Create a prompt with format instructions
prompt = PromptTemplate(
    template="Answer the user query.\n{format_instructions}\n{query}\n",
    input_variables=["query"],
    partial_variables={"format_instructions": parser.get_format_instructions()},
)

In [32]:
# Define a query to prompt the language model
query = "Tell me a joke."

In [33]:
# Combine prompt, model, and parser to get structured output
prompt_and_model = prompt | model
output = prompt_and_model.invoke({"query": query})

In [35]:
print(output)


{
    "setup": "Why did the tomato turn red?",
    "punchline": "Because it saw the salad dressing!"
}


In [36]:
# Parse the output using the parser
parsed_result = parser.invoke(output)

In [37]:
# The result is a structured object
print(parsed_result)

setup='Why did the tomato turn red?' punchline='Because it saw the salad dressing!'


#### <b><font size='3ptx'>SimpleJsonOutputParser</font></b>
Langchain's [**SimpleJsonOutputParser**](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.json.SimpleJsonOutputParser.html) is used when you want to parse JSON-like outputs. Here's an example:

In [38]:
from langchain.output_parsers.json import SimpleJsonOutputParser

# Create a JSON prompt
json_prompt = PromptTemplate.from_template(
    "Return a JSON object with `birthdate` and `birthplace` key that answers the following question: {question}"
)

In [42]:
# Initialize the JSON parser
json_parser = SimpleJsonOutputParser()

# Create a chain with the prompt, model, and parser
json_chain_no_parser = json_prompt | model
json_chain = json_prompt | model | json_parser

In [54]:
text_resp = json_chain_no_parser.invoke({"question": "When and where was Elon Musk born?"})

In [55]:
print(text_resp.__class__)
print(text_resp)

<class 'str'>


{
  "birthdate": "June 28, 1971",
  "birthplace": "Pretoria, South Africa"
}


In [52]:
# Stream through the results
json_resp = json_chain.invoke({"question": "When and where was Elon Musk born?"})

In [53]:
print(json_resp.__class__)
print(json_resp)

<class 'dict'>
{'birthdate': 'June 28, 1971', 'birthplace': 'Pretoria, South Africa'}


#### <b><font size='3ptx'>CommaSeparatedListOutputParser</font></b>
The [**CommaSeparatedListOutputParser**](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.list.CommaSeparatedListOutputParser.html) is handy when you want to extract comma-separated lists from model responses. Here's an example:

In [56]:
from langchain.output_parsers import CommaSeparatedListOutputParser
from langchain.prompts import PromptTemplate
from langchain.llms import OpenAI

# Initialize the parser
output_parser = CommaSeparatedListOutputParser()

# Create format instructions
format_instructions = output_parser.get_format_instructions()

# Create a prompt to request a list
prompt = PromptTemplate(
    template="List five {subject}.\n{format_instructions}",
    input_variables=["subject"],
    partial_variables={"format_instructions": format_instructions}
)

In [58]:
# Define a query to prompt the model
query = "English Premier League Teams"

# Generate the output
output = model.invoke(prompt.format(subject=query))

# Parse the output using the parser
parsed_result = output_parser.parse(output)

In [59]:
# The result is a list of items
print(parsed_result.__class__)
print(parsed_result)

<class 'list'>
['1. Manchester City', '2. Manchester United', '3. Liverpool', '4. Chelsea', '5. Arsenal']


These examples showcase how Langchain's output parsers can be used to structure various types of model responses, making them suitable for different applications and formats. Output parsers are a valuable tool for enhancing the usability and interpretability of language model outputs in Langchain.

<a id='sect_2'></a>
## <b><font color='darkblue'>Module II : Retrieval</font></b>
<b><font size='3ptx'>Retrieval in LangChain plays a crucial role in applications that require user-specific data, not included in the model's training set</font>. This process, known as [**Retrieval Augmented Generation**](https://python.langchain.com/docs/tutorials/rag/) (RAG), involves fetching external data and integrating it into the language model's generation process.</b>

<b>LangChain provides a comprehensive suite of tools and functionalities to facilitate this process</b>, catering to both simple and complex applications. LangChain achieves retrieval through a series of components which we will discuss one by one.

### <b><font color='darkgreen'>Document Loaders</font></b>
<b><font size='3ptx'>[Document loaders](https://python.langchain.com/docs/how_to/document_loader_custom/#overview) in LangChain enable the extraction of data from various sources</font>. With over 100 loaders available, they support a range of document types, apps and sources (private s3 buckets, public websites, databases)</b>.

You can choose a document loader based on your requirements [here](https://python.langchain.com/docs/integrations/document_loaders/).

<b>All these loaders ingest data into Document classes. We'll learn how to use data ingested into [Document](https://python.langchain.com/api_reference/core/documents/langchain_core.documents.base.Document.html) classes later.</b>

#### <b><font size='3ptx'>Text File Loader</font></b>
[**TextLoader**](https://python.langchain.com/api_reference/community/document_loaders/langchain_community.document_loaders.text.TextLoader.html) loads a simple `.txt` file into a document. e.g.:

In [61]:
from langchain.document_loaders import TextLoader

loader = TextLoader("test_data/sample.txt")
document = loader.load()

In [62]:
document

[Document(metadata={'source': 'test_data/sample.txt'}, page_content='Hello world! Hi John.\n')]

#### <b><font size='3ptx'>CSV Loader</fnot></b>
[**CSVLoader**](https://python.langchain.com/api_reference/community/document_loaders/langchain_community.document_loaders.csv_loader.CSVLoader.html) loads a CSV file into a document.

In [63]:
from langchain.document_loaders.csv_loader import CSVLoader

loader = CSVLoader(file_path='./test_data/sample.csv')
documents = loader.load()

In [64]:
documents

[Document(metadata={'source': './test_data/sample.csv', 'row': 0}, page_content='name: John\nage: 45'),
 Document(metadata={'source': './test_data/sample.csv', 'row': 1}, page_content='name: Mary\nage: 32'),
 Document(metadata={'source': './test_data/sample.csv', 'row': 2}, page_content='name: Peter\nage: 18')]

We can choose to customize the parsing by specifying field names

In [68]:
loader = CSVLoader(file_path='./test_data/sample2.csv', csv_args={
    'delimiter': ',',
    'quotechar': '"',
    'fieldnames': ['car brand', 'price']
})
documents = loader.load()

In [69]:
documents

[Document(metadata={'source': './test_data/sample2.csv', 'row': 0}, page_content='car brand: Toyota Camry\nprice: 25000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 1}, page_content='car brand: Honda Civic\nprice: 23000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 2}, page_content='car brand: Ford F-150\nprice: 40000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 3}, page_content='car brand: Chevrolet Silverado\nprice: 42000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 4}, page_content='car brand: BMW 3 Series\nprice: 45000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 5}, page_content='car brand: Mercedes-Benz C-Class\nprice: 48000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 6}, page_content='car brand: Tesla Model 3\nprice: 40000'),
 Document(metadata={'source': './test_data/sample2.csv', 'row': 7}, page_content='car brand: Hyundai Elantra\nprice: 21000'),

#### <b><font size='3ptx'>PDF Loaders</font></b>
PDF Loaders in LangChain offer various methods for parsing and extracting content from PDF files. Each loader caters to different requirements and uses different underlying libraries. [**PyPDFLoader**](https://python.langchain.com/api_reference/community/document_loaders/langchain_community.document_loaders.pdf.PyPDFLoader.html) is used for basic PDF parsing:

In [70]:
from langchain.document_loaders import PyPDFLoader

loader = PyPDFLoader("test_data/sample.pdf")
pages = loader.load_and_split()

In [71]:
pages

[Document(metadata={'source': 'test_data/sample.pdf', 'page': 0, 'page_label': '1'}, page_content='Hello  World!  Hi  John.')]

Let us now proceed and learn how to use these document classes.

### <b><font color='darkgreen'>Document Transformers</font></b>
<b><font size='3ptx'>[Document transformers](https://python.langchain.com/docs/integrations/document_transformers/) in LangChain are essential tools designed to manipulate documents, which we created in our previous subsection.</font>They are used for tasks such as splitting long documents into smaller chunks, combining, and filtering, which are crucial for adapting documents to a model's context window or meeting specific application needs.</b>

<b>One such tool is the [RecursiveCharacterTextSplitter](https://api.python.langchain.com/en/latest/character/langchain_text_splitters.character.RecursiveCharacterTextSplitter.html), a versatile text splitter that uses a character list for splitting</b>. It allows parameters like chunk size, overlap, and starting index. Here's an example of how it's used in Python:

In [101]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

# Initialize the language model
model = OpenAI(model_name="gpt-3.5-turbo-instruct", temperature=0.0)
long_text = model.invoke(
    'Please generate an article to introduce LLM with length just above 2000 characthers.'
    'The counting of character should exclude the Punctuation Marks.').strip()

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=100,
    chunk_overlap=20,
    length_function=len,
    add_start_index=True,
)

In [102]:
print(len(long_text))
print(long_text)

1356
The field of law is constantly evolving and becoming more complex, making it essential for legal professionals to stay updated and knowledgeable. This is where a Master of Laws (LLM) degree comes into play. An LLM is a postgraduate law degree that provides specialized legal education and training to individuals who have already completed their Juris Doctor (JD) or equivalent law degree.

The LLM degree is designed to enhance the legal skills and knowledge of practicing lawyers, law graduates, and professionals from other fields who wish to gain a deeper understanding of the law. It offers a wide range of specializations, including international law, corporate law, intellectual property law, human rights law, and more. This allows students to tailor their studies to their specific interests and career goals.

One of the main benefits of pursuing an LLM is the opportunity to gain a deeper understanding of a particular area of law. This is especially beneficial for lawyers who want t

In [103]:
texts = text_splitter.create_documents([long_text])
print(f'Total {len(texts)} documents generated!')

Total 18 documents generated!


In [104]:
print(texts[0])
print(texts[1])

page_content='The field of law is constantly evolving and becoming more complex, making it essential for legal' metadata={'start_index': 0}
page_content='essential for legal professionals to stay updated and knowledgeable. This is where a Master of Laws' metadata={'start_index': 77}


Another tool is the [**CharacterTextSplitter**](https://python.langchain.com/api_reference/text_splitters/character/langchain_text_splitters.character.CharacterTextSplitter.html), which splits text based on a specified character and includes controls for chunk size and overlap:

In [105]:
from langchain.text_splitter import CharacterTextSplitter

text_splitter = CharacterTextSplitter(
    separator="\n\n",
    chunk_size=500,
    chunk_overlap=50,
    length_function=len,
    is_separator_regex=False,
)

texts = text_splitter.create_documents([long_text])

In [106]:
print(texts[0])

page_content='The field of law is constantly evolving and becoming more complex, making it essential for legal professionals to stay updated and knowledgeable. This is where a Master of Laws (LLM) degree comes into play. An LLM is a postgraduate law degree that provides specialized legal education and training to individuals who have already completed their Juris Doctor (JD) or equivalent law degree.'


The [**HTMLHeaderTextSplitter**](https://python.langchain.com/api_reference/text_splitters/html/langchain_text_splitters.html.HTMLHeaderTextSplitter.html) is designed to split HTML content based on header tags, retaining the semantic structure:

In [111]:
# pip install lxml
!pip freeze | grep 'lxml'

lxml==5.3.0


In [118]:
from langchain.text_splitter import HTMLHeaderTextSplitter

html_content = model.invoke(
    'Please generate an article to introduce LLM with length just above 3000 characthers.'
    'Please use full HTML to wrap up this article and use header tags such as <h1>, <h2> etc to separate the sections.').strip()

In [122]:
Markdown(html_content)

<h1>Introducing the LLM Degree: A Comprehensive Guide</h1>

<p>The Master of Laws, or LLM, is a postgraduate degree that is highly sought after by law students and professionals around the world. It is a specialized program that allows students to deepen their knowledge and understanding of a specific area of law, and gain a competitive edge in the legal field. In this article, we will explore what the LLM degree is, its benefits, and how to pursue this prestigious qualification.</p>

<h2>What is an LLM Degree?</h2>

<p>The LLM degree is a postgraduate law degree that is typically pursued after completing a Bachelor of Laws (LLB) or Juris Doctor (JD) degree. It is an advanced program that focuses on a specific area of law, such as international law, corporate law, or intellectual property law. The duration of an LLM program can range from one to two years, depending on the country and university.</p>

<p>LLM programs are offered by top law schools around the world, and they attract students from diverse backgrounds, including recent law graduates, practicing lawyers, and professionals from other fields who wish to gain a deeper understanding of the law.</p>

<h2>Benefits

In [123]:
headers_to_split_on = [("h1", "Header 1"), ("h2", "Header 2")]

html_splitter = HTMLHeaderTextSplitter(headers_to_split_on=headers_to_split_on)
html_header_splits = html_splitter.split_text(html_content)

In [124]:
html_header_splits

[Document(metadata={'Header 1': 'Introducing the LLM Degree: A Comprehensive Guide'}, page_content='The Master of Laws, or LLM, is a postgraduate degree that is highly sought after by law students and professionals around the world. It is a specialized program that allows students to deepen their knowledge and understanding of a specific area of law, and gain a competitive edge in the legal field. In this article, we will explore what the LLM degree is, its benefits, and how to pursue this prestigious qualification.'),
 Document(metadata={'Header 1': 'Introducing the LLM Degree: A Comprehensive Guide', 'Header 2': 'What is an LLM Degree?'}, page_content='The LLM degree is a postgraduate law degree that is typically pursued after completing a Bachelor of Laws (LLB) or Juris Doctor (JD) degree. It is an advanced program that focuses on a specific area of law, such as international law, corporate law, or intellectual property law. The duration of an LLM program can range from one to two y

LangChain also offers specific splitters for different programming languages, like the Python Code Splitter and the JavaScript Code Splitter:

In [127]:
from langchain.text_splitter import RecursiveCharacterTextSplitter, Language

python_code = """
def hello_world():
    print("Hello, World!")
hello_world()
"""

python_splitter = RecursiveCharacterTextSplitter.from_language(
    language=Language.PYTHON, chunk_size=30, chunk_overlap=5
)
python_docs = python_splitter.create_documents([python_code])
print(python_docs[0])

js_code = """
function helloWorld() {
  console.log("Hello, World!");
}
helloWorld();
"""

js_splitter = RecursiveCharacterTextSplitter.from_language(
    language=Language.JS, chunk_size=30, chunk_overlap=5
)
js_docs = js_splitter.create_documents([js_code])
print(js_docs[0])

page_content='def hello_world():'
page_content='function helloWorld() {'


For splitting text based on token count, which is useful for language models with token limits, the [**TokenTextSplitter**](TokenTextSplitter) is used:

In [129]:
from langchain.text_splitter import TokenTextSplitter

text_splitter = TokenTextSplitter(chunk_size=100, chunk_overlap=20)
texts = text_splitter.split_text(long_text)
print(texts[0])

The field of law is constantly evolving and becoming more complex, making it essential for legal professionals to stay updated and knowledgeable. This is where a Master of Laws (LLM) degree comes into play. An LLM is a postgraduate law degree that provides specialized legal education and training to individuals who have already completed their Juris Doctor (JD) or equivalent law degree.

The LLM degree is designed to enhance the legal skills and knowledge of practicing lawyers, law graduates, and professionals from other


Finally, the [**LongContextReorder**](https://python.langchain.com/api_reference/community/document_transformers/langchain_community.document_transformers.long_context_reorder.LongContextReorder.html) reorders documents to prevent performance degradation in models due to long contexts:
```python
from langchain.document_transformers import LongContextReorder

reordering = LongContextReorder()
reordered_docs = reordering.transform_documents(docs)
print(reordered_docs[0])
```

These tools demonstrate various ways to transform documents in LangChain, from simple text splitting to complex reordering and language-specific splitting. For more in-depth and specific use cases, the LangChain documentation and Integrations section should be consulted.

In our examples, <b>the loaders have already created chunked documents for us, and this part is already handled.</b>

### <b><font color='darkgreen'>Text Embedding Models</font></b>
<b><font size='3ptx'>[Text embedding models](https://python.langchain.com/docs/how_to/embed_text/) in LangChain provide a standardized interface for various embedding model providers like OpenAI, Cohere, and Hugging Face</font>. These models transform text into vector representations, enabling operations like semantic search through text similarity in vector space.</b>

To get started with text embedding models, you typically need to install specific packages and set up API keys. We have already done this for OpenAI. 

In LangChain, the `embed_documents` method is used to embed multiple texts, providing a list of vector representations. For instance:

In [140]:
from langchain_openai import OpenAIEmbeddings

# Initialize the model
embeddings_model = OpenAIEmbeddings()

In [134]:
# Embed a list of texts
embeddings = embeddings_model.embed_documents(
    [
        "Hi there!",
        "Oh, hello!",
        "What's your name?",
        "My friends call me World",
        "Hello World!"
    ]
)

In [135]:
print("Number of documents embedded:", len(embeddings))
print("Dimension of each embedding:", len(embeddings[0]))

Number of documents embedded: 5
Dimension of each embedding: 1536


Let's try other embedding model provided by GCP Vertex API:

In [139]:
# TBD: ImportError: cannot import name 'replace_defs_in_schema'
#from langchain_google_vertexai import VertexAIEmbeddings
#embeddings_model = VertexAIEmbeddings(model="text-embedding-004")

For embedding a single text, such as a search query, the `embed_query` method is used. This is useful for comparing a query to a set of document embeddings. For example:

In [136]:
# Embed a single query
embedded_query = embeddings_model.embed_query("What was the name mentioned in the conversation?")
print("First five dimensions of the embedded query:", embedded_query[:5])

First five dimensions of the embedded query: [0.005329647101461887, -0.0006122003542259336, 0.0389961302280426, -0.002898985054343939, -0.008904732763767242]


Understanding these embeddings is crucial. Each piece of text is converted into a vector, the dimension of which depends on the model used. For instance, OpenAI models typically produce 1536-dimensional vectors. These embeddings are then used for retrieving relevant information.

LangChain's embedding functionality is not limited to OpenAI but is designed to work with various providers. The setup and usage might slightly differ depending on the provider, but the core concept of embedding texts into vector space remains the same. For detailed usage, including advanced configurations and integrations with different embedding model providers, the LangChain documentation in the Integrations section is a valuable resource.

### <b><font color='darkgreen'>Vector Stores</font></b>
<b><font size='3ptx'>[Vector stores](https://python.langchain.com/docs/integrations/vectorstores/) in LangChain support the efficient storage and searching of text embeddings</font>. LangChain integrates with over 50 vector stores, providing a standardized interface for ease of use.</b>

#### <b><font size='3ptx'>Example: Storing and Searching Embeddings</font></b>
After embedding texts, we can store them in a vector store like [**Chroma**](https://python.langchain.com/docs/integrations/vectorstores/chroma/) and perform similarity searches:

In [144]:
# pip install langchain_chroma
!pip freeze | grep 'langchain_chroma'

In [155]:
from langchain_chroma import Chroma
from langchain_openai import OpenAIEmbeddings

# Initialize the model
embeddings_model = OpenAIEmbeddings()

docs = [
    "The sun beat down mercilessly, making the asphalt shimmer with heat.",  # Weather
    "A sudden downpour drenched the city streets, sending pedestrians scurrying for cover.", # Weather
    "The gentle breeze carried the scent of rain and freshly cut grass.",  # Weather
    "The ripe mangoes dripped with sweet, golden juice.",  # Fruit
    "I love the tangy burst of flavor from a freshly squeezed lemon.",  # Fruit
    "The crisp apple provided a satisfying crunch with every bite.",  # Fruit
    "A majestic eagle soared high above the mountain peaks.",  # Animal
    "The playful kittens tumbled over each other, chasing a ball of yarn.",  # Animal
    "A lone wolf howled at the full moon, its mournful cry echoing through the valley.",  # Animal
    "The tiny hummingbird hovered delicately, sipping nectar from the vibrant flowers."  # Animal
]

text_splitter = CharacterTextSplitter(
    chunk_size=100,
    chunk_overlap=20,
    length_function=len,
    add_start_index=True,
)

# split it into chunks
texts = text_splitter.create_documents(docs)
texts

[Document(metadata={'start_index': 0}, page_content='The sun beat down mercilessly, making the asphalt shimmer with heat.'),
 Document(metadata={'start_index': 0}, page_content='A sudden downpour drenched the city streets, sending pedestrians scurrying for cover.'),
 Document(metadata={'start_index': 0}, page_content='The gentle breeze carried the scent of rain and freshly cut grass.'),
 Document(metadata={'start_index': 0}, page_content='The ripe mangoes dripped with sweet, golden juice.'),
 Document(metadata={'start_index': 0}, page_content='I love the tangy burst of flavor from a freshly squeezed lemon.'),
 Document(metadata={'start_index': 0}, page_content='The crisp apple provided a satisfying crunch with every bite.'),
 Document(metadata={'start_index': 0}, page_content='A majestic eagle soared high above the mountain peaks.'),
 Document(metadata={'start_index': 0}, page_content='The playful kittens tumbled over each other, chasing a ball of yarn.'),
 Document(metadata={'start_in

In [156]:
# vector_store = Chroma(embedding_function=embeddings_model)
db = Chroma.from_documents(texts, embeddings_model)

In [157]:
similar_texts = db.similarity_search("weather", k=3)

In [159]:
for text in similar_texts:
    print(text)

page_content='A sudden downpour drenched the city streets, sending pedestrians scurrying for cover.' metadata={'start_index': 0}
page_content='The sun beat down mercilessly, making the asphalt shimmer with heat.' metadata={'start_index': 0}
page_content='The gentle breeze carried the scent of rain and freshly cut grass.' metadata={'start_index': 0}


Let us alternatively use the [**FAISS vector store**](https://python.langchain.com/docs/integrations/vectorstores/faiss/) to create indexes for our documents:

In [161]:
from langchain.vectorstores import FAISS

faiss_vs = FAISS.from_documents(texts, embedding=embeddings_model)

In [163]:
similar_texts = faiss_vs.similarity_search("fruit", k=3)

In [164]:
for text in similar_texts:
    print(text)

page_content='The ripe mangoes dripped with sweet, golden juice.' metadata={'start_index': 0}
page_content='The crisp apple provided a satisfying crunch with every bite.' metadata={'start_index': 0}
page_content='I love the tangy burst of flavor from a freshly squeezed lemon.' metadata={'start_index': 0}


### <b><font color='darkgreen'>Retrievers</font></b>
<b><font size='3ptx'>[Retrievers](https://python.langchain.com/docs/concepts/retrievers/) in LangChain are interfaces that return documents in response to an unstructured query</font>. They are more general than vector stores, focusing on retrieval rather than storage. Although vector stores can be used as a retriever's backbone, there are other types of retrievers as well.</b>
![flow](https://python.langchain.com/assets/images/retriever_concept-1093f15a8f63ddb90bd23decbd249ea5.png)

To set up a Chroma retriever, you first install it using
```shell
$ pip install chromadb
```
. Then, you load, split, embed, and retrieve documents using a series of Python commands. Here's a code example for setting up a Chroma retriever:

In [176]:
full_text = open("test_data/CockatooAI_README.md", "r").read()
text_splitter = CharacterTextSplitter(chunk_size=500, chunk_overlap=50)
texts = text_splitter.split_text(full_text)

Created a chunk of size 1161, which is longer than the specified 500
Created a chunk of size 579, which is longer than the specified 500
Created a chunk of size 587, which is longer than the specified 500
Created a chunk of size 1142, which is longer than the specified 500


In [177]:
embeddings = OpenAIEmbeddings()
db = Chroma.from_texts(texts, embeddings)
retriever = db.as_retriever()

In [178]:
retrieved_docs = retriever.invoke("What does model A mean?")

In [179]:
print(f'Total {len(retrieved_docs)} docs obtained!')

Total 4 docs obtained!


In [180]:
print(retrieved_docs[0].page_content)

# How To (Initiative):
Use models (A-B-C as shown below) to form the pipeline as the final language tutor:
- model A: voice to text
- model B(LLM): text à text , might be managed by [langChain](https://python.langchain.com/docs/get_started/introduction), Andrew Ng provided [2-3 free short courses](https://www.deeplearning.ai/short-courses/).
- model C: text to voice
PS: C might be the same as A, needs more research. But the** Goal towards users** should be the top concern when considering which model to use. Besides, better to use open-source models instead of paying premium APIs.


<b>The [**MultiQueryRetriever**](https://python.langchain.com/docs/how_to/MultiQueryRetriever/) automates the process of prompt tuning by using an LLM to generate multiple queries from different perspectives for a given user input query</b>. For each query, it retrieves a set of relevant documents and takes the unique union across all queries to get a larger set of potentially relevant documents. By generating multiple perspectives on the same question, the MultiQueryRetriever can mitigate some of the limitations of the distance-based retrieval and get a richer set of results.

In [182]:
from langchain_openai import ChatOpenAI
from langchain.retrievers.multi_query import MultiQueryRetriever

question = "List the three models used in Cockatoo AI and give a short description to them."
llm = ChatOpenAI(temperature=0)
retriever_from_llm = MultiQueryRetriever.from_llm(
    retriever=db.as_retriever(), llm=llm
)

In [185]:
unique_docs = retriever_from_llm.invoke(input=question)
print("Number of unique documents:", len(unique_docs))

Number of unique documents: 4


In [190]:
print(unique_docs[0].page_content)

# Project language tutor (focused on listening & speaking)
Cockatoo AI is a cutting-edge language tutor project focused on enhancing listening and speaking skills.  Our AI companion engages users in realistic conversations, adapting its speaking style (from direct to philosophical) and difficulty level (A1-C1) to suit individual learning needs. Cockatoo AI can discuss user-provided topics, ask relevant questions, and even express itself with a variety of tones and voices (happy, sad, elderly, adult, male, female, etc.) to provide comprehensive listening practice.  For our team, this project provides valuable experience in LLM development, voice model integration, deep learning, and CI/CD, while also offering opportunities to explore bias mitigation techniques and optimize the balance between fine-tuning and computational resources.  Our architecture utilizes a three-stage pipeline: voice-to-text (Model A), text-to-text processing with LLMs (Model B, potentially managed by LangChain), a

<b>[Contextual Compression](https://python.langchain.com/docs/how_to/contextual_compression/) in LangChain compresses retrieved documents using the context of the query, ensuring only relevant information is returned</b>. This involves content reduction and filtering out less relevant documents. The following code example shows how to use Contextual Compression Retriever:

In [189]:
from langchain_openai import ChatOpenAI
from langchain.retrievers import ContextualCompressionRetriever
from langchain.retrievers.document_compressors import LLMChainExtractor

llm = OpenAI(temperature=0)
compressor = LLMChainExtractor.from_llm(llm)
compression_retriever = ContextualCompressionRetriever(base_compressor=compressor, base_retriever=retriever)

In [191]:
question = "List the three models used in Cockatoo AI and give a short description to them."
compressed_docs = compression_retriever.get_relevant_documents(question)

In [193]:
print(f'Total {len(compressed_docs)} docs obtained:')
print(compressed_docs[0].page_content)

Total 4 docs obtained:
- Cockatoo AI is a cutting-edge language tutor project focused on enhancing listening and speaking skills.
- Our AI companion engages users in realistic conversations, adapting its speaking style (from direct to philosophical) and difficulty level (A1-C1) to suit individual learning needs.
- Cockatoo AI can discuss user-provided topics, ask relevant questions, and even express itself with a variety of tones and voices (happy, sad, elderly, adult, male, female, etc.) to provide comprehensive listening practice.
- Our architecture utilizes a three-stage pipeline: voice-to-text (Model A), text-to-text processing with LLMs (Model B, potentially managed by LangChain), and text-to-voice (Model C, possibly the same as A).
- We prioritize open-source models and aim to create a personalized and effective language learning experience.


The [**EnsembleRetriever**](https://api.python.langchain.com/en/latest/retrievers/langchain.retrievers.ensemble.EnsembleRetriever.html) combines different retrieval algorithms to achieve better performance. An example of combining BM25 and FAISS Retrievers is shown in the following code ([more](https://python.langchain.com/docs/how_to/ensemble_retriever/)):

In [255]:
#pip install rank_bm25
#pip install lark
!pip freeze | grep -P '(bm25|lark)'

lark==1.2.2
rank-bm25==0.2.2


In [208]:
texts[0]

'# Project language tutor (focused on listening & speaking)\nCockatoo AI is a cutting-edge language tutor project focused on enhancing listening and speaking skills.  Our AI companion engages users in realistic conversations, adapting its speaking style (from direct to philosophical) and difficulty level (A1-C1) to suit individual learning needs. Cockatoo AI can discuss user-provided topics, ask relevant questions, and even express itself with a variety of tones and voices (happy, sad, elderly, adult, male, female, etc.) to provide comprehensive listening practice.  For our team, this project provides valuable experience in LLM development, voice model integration, deep learning, and CI/CD, while also offering opportunities to explore bias mitigation techniques and optimize the balance between fine-tuning and computational resources.  Our architecture utilizes a three-stage pipeline: voice-to-text (Model A), text-to-text processing with LLMs (Model B, potentially managed by LangChain),

In [209]:
from langchain.retrievers import BM25Retriever, EnsembleRetriever
from langchain.vectorstores import FAISS

bm25_retriever = BM25Retriever.from_texts(texts, k=2)
faiss_vs = FAISS.from_texts(texts, embedding=embeddings_model)
faiss_retriever = faiss_vs.as_retriever(search_kwargs={"k": 2})

ensemble_retriever = EnsembleRetriever(
    retrievers=[bm25_retriever, faiss_retriever], weights=[0.5, 0.5])

In [211]:
docs = ensemble_retriever.get_relevant_documents("model A")

In [213]:
print(f'Total {len(docs)} docs obtained!')
print(docs[1].page_content)

Total 3 docs obtained!
# Good References
## For Model A (Speech to text, STT)
- The library [`SpeechRecognition`](https://pypi.org/project/SpeechRecognition/)
  is used for performing speech recognition, with support for several engines and APIs, online and offline.
    - [Notebook - Easy Speech-to-Text with Python](https://github.com/johnklee/ml_articles/blob/master/medium/Easy-speech-to-text-with-python/notebook.ipynb)
- [`Whisper`](https://github.com/openai/whisper) is a general-purpose speech recognition model. It is trained on a large dataset of diverse audio and is also a multitasking model that can perform multilingual speech recognition, speech translation, and language identification.
    - [Notebook - Introduction for Whisper in OpenAI](https://github.com/johnklee/ml_articles/blob/master/others/ithome_ithelp_openapi_whisper/notebook.ipynb)
## For Model C (Text to speech, TTS)
- The company [Evelenlabs](https://elevenlabs.io/) has advanced **non-open sourced** TTS model for mu

Generating summaries for better retrieval due to more focused content representation is another method. Here's an example of generating summaries:

In [227]:
from langchain_openai import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain.retrievers.multi_vector import MultiVectorRetriever
from langchain.schema.output_parser import StrOutputParser
from langchain.schema.document import Document
from langchain.storage import InMemoryStore
import uuid


vectorstore = Chroma(collection_name="full_documents", embedding_function=OpenAIEmbeddings())
store = InMemoryStore()
id_key = "doc_id"
retriever = MultiVectorRetriever(vectorstore=vectorstore, docstore=store, id_key=id_key)
doc_ids = [str(uuid.uuid4()) for _ in docs]

chain = (
    (lambda x: x.page_content) |
    ChatPromptTemplate.from_template("Summarize the following document:\n\n{doc}") |
    ChatOpenAI(max_retries=0) |
    StrOutputParser())
summaries = chain.batch(docs, {"max_concurrency": 5})

In [228]:
summaries

['The document is titled "Misc" and likely contains miscellaneous information or content. It is not clear what specific information is included in the document based on the title alone.',
 'The document provides good references for speech to text (STT) and text to speech (TTS) models. It mentions the `SpeechRecognition` library for STT, with various engines and APIs, and the `Whisper` model for general-purpose speech recognition. For TTS, it recommends the non-open sourced TTS model by Evelenlabs, which excels in multi-languages and various tones and sounds. Links to notebooks for each model are also provided for further exploration.',
 'The document outlines a method for creating a language tutor using three models in a pipeline: voice to text (model A), text to text (model B), and text to voice (model C). It suggests using open-source models rather than paid APIs and emphasizes prioritizing the goal towards users when selecting which model to use. Research is needed to determine if m

In [229]:
summary_docs = [
    Document(page_content=s, metadata={id_key: doc_ids[i]}) for i, s in enumerate(summaries)]

In [230]:
retriever.vectorstore.add_documents(summary_docs)
retriever.docstore.mset(list(zip(doc_ids, docs)))

A self-querying retriever [**SelfQueryRetriever**](https://api.python.langchain.com/en/latest/retrievers/langchain.retrievers.self_query.base.SelfQueryRetriever.html) is one that, as the name suggests, has the ability to query itself. Specifically, <b>given any natural language query, the retriever uses a query-constructing LLM chain to write a structured query and then applies that structured query to its underlying vector store</b>. This allows the retriever to not only use the user-input query for semantic similarity comparison with the contents of stored documents but to also extract filters from the user query on the metadata of stored documents and to execute those filters ([details](https://python.langchain.com/docs/how_to/self_query/)).
![flow](https://python.langchain.com/assets/images/self_querying-26ac0fc8692e85bc3cd9b8640509404f.jpg)

In [20]:
from langchain_openai import ChatOpenAI
from langchain_chroma import Chroma
from langchain_core.documents import Document
from langchain.chains.query_constructor.base import AttributeInfo
from langchain.retrievers.self_query.base import SelfQueryRetriever
from langchain_openai import OpenAIEmbeddings

docs = [
    Document(
        page_content="A bunch of scientists bring back dinosaurs and mayhem breaks loose",
        metadata={"year": 1993, "rating": 7.7, "genre": "science fiction"},
    ),
    Document(
        page_content="Leo DiCaprio gets lost in a dream within a dream within a dream within a ...",
        metadata={"year": 2010, "director": "Christopher Nolan", "rating": 8.2},
    ),
    Document(
        page_content="A psychologist / detective gets lost in a series of dreams within dreams within dreams and Inception reused the idea",
        metadata={"year": 2006, "director": "Satoshi Kon", "rating": 8.6},
    ),
    Document(
        page_content="A bunch of normal-sized women are supremely wholesome and some men pine after them",
        metadata={"year": 2019, "director": "Greta Gerwig", "rating": 8.3},
    ),
    Document(
        page_content="Toys come alive and have a blast doing so",
        metadata={"year": 1995, "genre": "animated"},
    ),
    Document(
        page_content="Three men walk into the Zone, three men walk out of the Zone",
        metadata={
            "year": 1979,
            "director": "Andrei Tarkovsky",
            "genre": "thriller",
            "rating": 9.9,
        },
    ),
]
vectorstore = Chroma.from_documents(docs, OpenAIEmbeddings())

In [21]:
metadata_field_info = [
    AttributeInfo(
        name="genre",
        description="The genre of the movie. One of ['science fiction', 'comedy', 'drama', 'thriller', 'romance', 'action', 'animated']",
        type="string",
    ),
    AttributeInfo(
        name="year",
        description="The year the movie was released",
        type="integer",
    ),
    AttributeInfo(
        name="director",
        description="The name of the movie director",
        type="string",
    ),
    AttributeInfo(
        name="rating", description="A 1-10 rating for the movie", type="float"
    ),
]
document_content_description = "Brief summary of a movie"
llm = ChatOpenAI(temperature=0)
retriever = SelfQueryRetriever.from_llm(
    llm,
    vectorstore,
    document_content_description,
    metadata_field_info,
)

In [10]:
retriever = SelfQueryRetriever.from_llm(
    llm, vectorstore, document_content_description, metadata_field_info)

In [24]:
# This example only specifies a filter
retrieved_docs = retriever.invoke("I want to watch a movie rated higher than 8.5")

In [25]:
print(f'Total {len(retrieved_docs)} docs obtained!')

Total 2 docs obtained!


In [29]:
for i, doc in enumerate(retrieved_docs):
    print(f'{i+1}. {doc.page_content}')

1. Three men walk into the Zone, three men walk out of the Zone
2. A psychologist / detective gets lost in a series of dreams within dreams within dreams and Inception reused the idea


<a id='sect_3'></a>
## <b><font color='darkblue'>Module III : Agents</font></b>
<b><font size='3ptx'>LangChain introduces a powerful concept called "[Agents](https://python.langchain.com/docs/tutorials/agents/)" that takes the idea of chains to a whole new level</font>. Agents leverage language models to dynamically determine sequences of actions to perform, making them incredibly versatile and adaptive. Unlike traditional chains, where actions are hardcoded in code, agents employ language models as reasoning engines to decide which actions to take and in what order</b>.

**The Agent is the core component responsible for decision-making**. It harnesses the power of a language model and a prompt to determine the next steps to achieve a specific objective. The inputs to an agent typically include:
* **Tools**: Descriptions of available tools (more on this later).
* **User Input**: The high-level objective or query from the user.
* **Intermediate Steps**: A history of (action, tool output) pairs executed to reach the current user input.

The output of an agent can be the next action to take actions ([**AgentActions**](https://api.python.langchain.com/en/latest/agents/langchain_core.agents.AgentAction.html)) or the final response to send to the user ([**AgentFinish**](https://api.python.langchain.com/en/latest/agents/langchain_core.agents.AgentFinish.html)). An action specifies a tool and the input for that tool.

### <b><font color='darkgreen'>Tools</font></b>
<b><font size='3ptx'>[Tools](https://python.langchain.com/docs/tutorials/agents/#define-tools) are interfaces that an agent can use to interact with the world</font>. They enable agents to perform various tasks, such as searching the web, running shell commands, or accessing external APIs. In LangChain, tools are essential for extending the capabilities of agents and enabling them to accomplish diverse tasks</b>.

To use tools in LangChain, you can load them using the following snippet:
```python
from langchain.agents import load_tools

tool_names = [...]
tools = load_tools(tool_names)
```

Some tools may require a base Language Model (LLM) to initialize. In such cases, you can pass an LLM as well:
```python
from langchain.agents import load_tools

tool_names = [...]
llm = ...
tools = load_tools(tool_names, llm=llm)
```

This setup allows you to access a variety of tools and integrate them into your agent's workflows. The complete list of tools with usage documentation is [here](https://python.langchain.com/docs/integrations/tools/).

#### <b><font size='3ptx'>DuckDuckGo</font></b>
The [**DuckDuckGo**](https://python.langchain.com/docs/integrations/tools/ddg/) tool enables you to perform web searches using its search engine. Here's how to use it:

In [32]:
from langchain_community.tools import DuckDuckGoSearchRun

# search = DuckDuckGoSearchRun()
# search.invoke("Obama's first name?")

#### <b><font size>Shell (bash)</font></b>
The Shell toolkit provides agents with access to the shell environment, allowing them to execute shell commands. This feature is powerful but should be used with caution, especially in sandboxed environments. Here's how you can use the Shell tool:

In [33]:
from langchain.tools import ShellTool

shell_tool = ShellTool()

result = shell_tool.run({"commands": ["echo 'Hello World!'", "time"]})

Executing command:
 ["echo 'Hello World!'", 'time']


/usr/local/google/home/johnkclee/Github/ml_articles/env/lib/python3.11/site-packages/langchain_community/tools/shell/tool.py:33: UserWarning: The shell tool has no safeguards by default. Use at your own risk.
  warnings.warn(


In [35]:
print(result)

Hello World!
user	0m0.00s
sys	0m0.00s



### <b><font color='darkgreen'>Back to Agents</font></b>
Let's move on to agents now... TBD

<a id='sect_4'></a>
## <b><font color='darkblue'>Module IV : Chains</font></b> ([back](#agenda))
<b><font size='3ptx'>LangChain is a tool designed for utilizing Large Language Models (LLMs) in complex applications</font></b>. It provides frameworks for creating chains of components, including LLMs and other types of components. Two primary frameworks:
* The LangChain Expression Language (LCEL)
* Legacy Chain interface

<b>[The LangChain Expression Language](https://python.langchain.com/docs/concepts/lcel/) (LCEL) is a syntax that allows for intuitive composition of chains</b>. It supports advanced features like streaming, asynchronous calls, batching, parallelization, retries, fallbacks, and tracing.

For example, you can compose a prompt, model, and output parser in LCEL as shown in the following code:

In [37]:
from langchain.prompts import ChatPromptTemplate
from langchain.schema import StrOutputParser

model = ChatOpenAI(model="gpt-3.5-turbo", temperature=0)
prompt = ChatPromptTemplate.from_messages([
    ("system", "You're a very knowledgeable historian who provides accurate and eloquent answers to historical questions."),
    ("human", "{question}")
])
runnable = prompt | model | StrOutputParser()

In [38]:
resp = runnable.invoke({"question": "What are the seven wonders of the world"})

In [40]:
Markdown(resp)

The Seven Wonders of the Ancient World were a list of remarkable constructions of classical antiquity. They are:

1. The Great Pyramid of Giza, Egypt
2. The Hanging Gardens of Babylon, Iraq
3. The Statue of Zeus at Olympia, Greece
4. The Temple of Artemis at Ephesus, Turkey
5. The Mausoleum at Halicarnassus, Turkey
6. The Colossus of Rhodes, Greece
7. The Lighthouse of Alexandria, Egypt

It's worth noting that only the Great Pyramid of Giza still exists today, as the other wonders have been destroyed over time.

Alternatively, the [**LLMChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.llm.LLMChain.html#llmchain) is an option similar to LCEL for composing components. The [**LLMChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.llm.LLMChain.html#llmchain) example is as follows:

In [43]:
from langchain.chains import LLMChain

chain = LLMChain(llm=model, prompt=prompt, output_parser=StrOutputParser())
resp = chain.invoke(input="What are the seven wonders of the world")

In [46]:
resp

{'question': 'What are the seven wonders of the world',
 'text': 'The Seven Wonders of the Ancient World were a list of remarkable constructions of classical antiquity. They included the Great Pyramid of Giza, the Hanging Gardens of Babylon, the Statue of Zeus at Olympia, the Temple of Artemis at Ephesus, the Mausoleum at Halicarnassus, the Colossus of Rhodes, and the Lighthouse of Alexandria. These wonders were considered marvels of architecture and engineering at the time and have captured the imagination of people for centuries.'}

Chains in [**LLMChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.llm.LLMChain.html#llmchain) can also be stateful by incorporating a Memory object ([**ConversationChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.conversation.base.ConversationChain.html#conversationchain): Chain to have a conversation and load context from memory.). This allows for data persistence across calls, as shown in this example:

In [56]:
from langchain.chains import ConversationChain
from langchain.memory import ConversationBufferMemory
from langchain_core.chat_history import InMemoryChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory
from langchain_google_genai import ChatGoogleGenerativeAI

store = {}  # memory is maintained outside the chain

def get_session_history(session_id: str) -> InMemoryChatMessageHistory:
    if session_id not in store:
        store[session_id] = InMemoryChatMessageHistory()
    return store[session_id]

config = {"configurable": {"session_id": "1"}}
chat = ChatGoogleGenerativeAI(
    model="gemini-1.5-pro",
    temperature=0,
    max_tokens=None,
    timeout=None,
    max_retries=2,
    # other params...
)
chain = RunnableWithMessageHistory(chat, get_session_history)
#conversation = ConversationChain(llm=chat, memory=ConversationBufferMemory())

In [57]:
resp = chain.invoke(
    "Answer briefly. What are the first 3 colors of a rainbow?", config)

In [59]:
print(f'{resp.__class__}')
print(resp.content)

<class 'langchain_core.messages.ai.AIMessage'>
Red, orange, yellow


In [60]:
resp = chain.invoke("And the next 4?", config)

In [61]:
print(f'{resp.__class__}')
print(resp.content)

<class 'langchain_core.messages.ai.AIMessage'>
Green, blue, indigo, violet


<b>LangChain also supports integration with OpenAI's function-calling APIs, which is useful for obtaining structured outputs and executing functions within a chain</b>. For getting structured outputs, you can specify them using Pydantic classes or JsonSchema, as illustrated below:

In [68]:
from typing import Optional
from pydantic import BaseModel, Field, field_validator
from langchain.chains.openai_functions import create_structured_output_runnable
from langchain_openai import ChatOpenAI
from langchain.prompts import ChatPromptTemplate

class Person(BaseModel):
    name: str = Field(description="The person's name")
    age: int = Field(description="The person's age")
    fav_food: Optional[str] = Field(None, description="The person's favorite food")


llm = ChatOpenAI(model="gpt-4", temperature=0)
prompt = ChatPromptTemplate.from_messages([
    # Prompt messages here
])
runnable = create_structured_output_runnable(Person, llm, prompt)

In [70]:
resp = runnable.invoke("Sally is 13")

In [71]:
resp

Person(name='Sally', age=13, fav_food=None)

<b>LangChain leverages OpenAI functions to create various specific chains for different purposes</b>. These include chains for extraction, tagging, OpenAPI, and QA with citations.

<b>In the context of extraction, the process is similar to the structured output chain but focuses on information or entity extraction</b> ([more](https://python.langchain.com/docs/how_to/structured_output/)). For tagging, the idea is to label a document with classes such as sentiment, language, style, covered topics, or political tendency.

An example of how tagging works in LangChain can be demonstrated with a Python code. The process begins with installing the necessary packages and setting up the environment:

In [72]:
from langchain_openai import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain.chains import create_tagging_chain, create_tagging_chain_pydantic

The schema for tagging is defined, specifying the properties and their expected types:

In [79]:
schema = {
    "properties": {
        "sentiment": {"type": "string"},
        "aggressiveness": {"type": "integer"},
        "language": {"type": "string"},
    }
}

class Schema(BaseModel):
    sentiment: str = Field(description="The sentiment of sentences")
    aggressiveness: int = Field(description="The aggressiveness of sentence")
    language: str = Field(None, description="The spoken language")

#llm = ChatOpenAI(temperature=0, model="gpt-3.5-turbo-0613")
#chain = create_tagging_chain(schema, llm)

In [80]:
import getpass
import os

if not os.environ.get("OPENAI_API_KEY"):
  os.environ["OPENAI_API_KEY"] = getpass.getpass("Enter API key for OpenAI: ")

llm = ChatOpenAI(model="gpt-4o-mini")
structured_llm = llm.with_structured_output(Schema)

In [81]:
resp = structured_llm.invoke("Estoy increiblemente contento de haberte conocido! Creo que seremos muy buenos amigos!")

In [82]:
resp

Schema(sentiment='positivo', aggressiveness=0, language='español')

In [83]:
resp = structured_llm.invoke("This movie really sucks and I won't recommend it!")

In [84]:
resp

Schema(sentiment='Negative', aggressiveness=7, language='English')

<b>Additionally, LangChain's metadata tagger document transformer can be used to extract metadata from LangChain Documents</b>, offering similar functionality to the tagging chain but applied to a LangChain Document.

Citing retrieval sources is another feature of LangChain, using OpenAI functions to extract citations from text. This is demonstrated in the following code:

In [87]:
from langchain.chains import create_citation_fuzzy_match_runnable
from langchain_openai import ChatOpenAI

llm = ChatOpenAI(model="gpt-4o-mini")

context = "Alice has blue eyes. Bob has brown eyes. Charlie has green eyes."
question = "What color are Bob's eyes?"

chain = create_citation_fuzzy_match_runnable(llm)

In [88]:
chain.invoke({"question": question, "context": context})

QuestionAnswer(question="What color are Bob's eyes?", answer=[FactWithEvidence(fact='Bob has brown eyes.', substring_quote=['Bob has brown eyes'])])

<b>In LangChain, chaining in Large Language Model (LLM) applications typically involves combining a prompt template with an LLM and optionally an output parser.</b> The recommended way to do this is through the LangChain Expression Language (LCEL), although the legacy [**LLMChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.llm.LLMChain.html#llmchain) approach is also supported.

Using LCEL, the [**BasePromptTemplate**](https://python.langchain.com/api_reference/core/prompts/langchain_core.prompts.base.BasePromptTemplate.html), [**BaseLanguageModel**](https://python.langchain.com/api_reference/core/language_models/langchain_core.language_models.base.BaseLanguageModel.html), and [**BaseOutputParser**](https://python.langchain.com/api_reference/core/output_parsers/langchain_core.output_parsers.base.BaseOutputParser.html) all implement the [**Runnable**](https://python.langchain.com/api_reference/core/runnables/langchain_core.runnables.base.Runnable.html#langchain_core.runnables.base.Runnable) interface and can be easily piped into one another. Here's an example demonstrating this:

In [91]:
from langchain.prompts import PromptTemplate
from langchain_openai import ChatOpenAI
from langchain.schema import StrOutputParser

prompt = PromptTemplate.from_template(
    "What is a good name for a company that makes {product}?"
)
runnable = prompt | ChatOpenAI() | StrOutputParser()

In [92]:
runnable.invoke({"product": "colorful socks"})

'RainbowSock Co.'

<b>Routing in LangChain allows for creating non-deterministic chains where the output of a previous step determines the next step</b>. This helps in structuring and maintaining consistency in interactions with LLMs. <b>For instance, if you have two templates optimized for different types of questions, you can choose the template based on user input</b>.

Here's how you can achieve this using LCEL with a [**RunnableBranch**](https://python.langchain.com/api_reference/core/runnables/langchain_core.runnables.branch.RunnableBranch.html#langchain_core.runnables.branch.RunnableBranch), which is initialized with a list of (condition, runnable) pairs and a default runnable:

In [126]:
from operator import itemgetter
from langchain_openai import ChatOpenAI
from langchain.schema.output_parser import StrOutputParser
from langchain.schema.runnable import RunnableBranch
from langchain.schema.runnable import RunnablePassthrough

class Topic(BaseModel):
    input: str = Field(description="The original question.")
    topic: str = Field(description='topic name (e.g., "math", "physics", or "general")')
    

def output_parser_to_dict(topic_object: Topic) -> dict:
    """Parses the Topic object and returns a dictionary."""
    return {'input': topic_object.input, 'topic': topic_object.topic}
        

question = "What is the first prime number greater than 40 and it is divisible by 3?"

general_prompt = PromptTemplate.from_template(
    "You are a helpful assistant. Answer the question as accurately as you can.\n\n{input}"
)

math_prompt = PromptTemplate.from_template(
    "You are a mathematician . Answer the question as easy as you can.\n\n{input}"
)

physics_prompt = PromptTemplate.from_template(
    "You are a smart physicist. Answer the question as intuitive as you can.\n\n{input}"
)

prompt_branch = RunnableBranch(
    (lambda x: x["topic"] == "math", math_prompt),
    (lambda x: x["topic"] == "physics", physics_prompt),
    general_prompt,
)

classifier_prompt = PromptTemplate.from_template(
    """Classify the following question into one of the following topics: math, physics, or general. 
    Respond with only the topic name (e.g., "math", "physics", or "general").

    Question: {input}
    """
)

structured_llm = ChatOpenAI().with_structured_output(Topic)
classifier_chain = classifier_prompt | structured_llm | output_parser_to_dict

/usr/local/google/home/johnkclee/Github/ml_articles/env/lib/python3.11/site-packages/langchain_openai/chat_models/base.py:1375: UserWarning: Cannot use method='json_schema' with model gpt-3.5-turbo since it doesn't support OpenAI's Structured Output API. You can see supported models here: https://platform.openai.com/docs/guides/structured-outputs#supported-models. To fix this warning, set `method='function_calling'. Overriding to method='function_calling'.
  warnings.warn(


In [127]:
classifier_chain.invoke(question)

{'input': 'What is the first prime number greater than 40 and it is divisible by 3?',
 'topic': 'math'}

The final chain is then constructed using various components, such as a topic classifier, prompt branch, and an output parser, to determine the flow based on the topic of the input:

In [121]:
final_chain = (
    classifier_chain | RunnablePassthrough.assign(topic=itemgetter("topic"))
    | prompt_branch
    | ChatOpenAI()
    | StrOutputParser()
)

In [124]:
resp = final_chain.invoke("What is the first prime number greater than 40 and it is divisible by 3?")

In [125]:
resp

'The first prime number greater than 40 that is divisible by 3 is 43.'

This approach exemplifies the flexibility and power of LangChain in handling complex queries and routing them appropriately based on the input.

<b>In the realm of language models, a common practice is to follow up an initial call with a series of subsequent calls, using the output of one call as input for the next</b>. This sequential approach is especially beneficial when you want to build on the information generated in previous interactions. While <b>[the LangChain Expression Language](https://python.langchain.com/docs/concepts/lcel/) (LCEL) is the recommended method for creating these sequences</b>, the [**SequentialChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.sequential.SequentialChain.html) method is still documented for its backward compatibility.

To illustrate this, let's consider a scenario where we first generate a play synopsis and then a review based on that synopsis. Using Python's
`langchain.prompts` , we create two [**PromptTemplate**](https://python.langchain.com/api_reference/core/prompts/langchain_core.prompts.prompt.PromptTemplate.html#langchain_core.prompts.prompt.PromptTemplate) instances: one for the synopsis and another for the review. Here's the code to set up these templates:

In [128]:
from langchain.prompts import PromptTemplate


synopsis_prompt = PromptTemplate.from_template(
    "You are a playwright. Given the title of play, it is your job to write a synopsis for that title.\n\nTitle: {title}\nPlaywright: This is a synopsis for the above play:"
)

review_prompt = PromptTemplate.from_template(
    "You are a play critic from the New York Times. Given the synopsis of play, it is your job to write a review for that play.\n\nPlay Synopsis:\n{synopsis}\nReview from a New York Times play critic of the above play:"
)

In the LCEL approach, we chain these prompts with `ChatOpenAI` and `StrOutputParser` to create a sequence that first generates a synopsis and then a review. The code snippet is as follows:

In [129]:
from langchain_openai import ChatOpenAI
from langchain.schema import StrOutputParser

llm = ChatOpenAI()
chain = (
    {"synopsis": synopsis_prompt | llm | StrOutputParser()}
    | review_prompt
    | llm
    | StrOutputParser()
)

In [130]:
resp = chain.invoke({"title": "Tragedy at sunset on the beach"})

In [132]:
print(resp)

"Tragedy at Sunset on the Beach" is a poignant and gripping production that will leave audiences emotionally moved and on the edge of their seats. Set against the backdrop of a picturesque beach at dusk, the play beautifully captures the complexities of love, loss, and betrayal.

The chemistry between the two leads, Sarah and Jack, is palpable and their performances are raw and heartfelt. As the dark secret from the past unravels, the tension between them escalates, keeping the audience engaged and invested in their story.

The tragedy that unfolds is heartbreaking and beautifully executed, leaving a lasting impact on the audience. The skillful direction and powerful performances bring the emotional depth of the story to life, leaving no dry eye in the theater.

In the end, "Tragedy at Sunset on the Beach" is a must-see for theatergoers looking for a thought-provoking and emotionally resonant production. Love may be tested, but the power of the human spirit shines through in this unfor

If we need both the synopsis and the review, we can use [**RunnablePassthrough**](https://python.langchain.com/api_reference/core/runnables/langchain_core.runnables.passthrough.RunnablePassthrough.html#langchain_core.runnables.passthrough.RunnablePassthrough) to create a separate chain for each and then combine them:

In [133]:
from langchain.schema.runnable import RunnablePassthrough

synopsis_chain = synopsis_prompt | llm | StrOutputParser()
review_chain = review_prompt | llm | StrOutputParser()
chain = {"synopsis": synopsis_chain} | RunnablePassthrough.assign(review=review_chain)

In [134]:
resp = chain.invoke({"title": "Tragedy at sunset on the beach"})

In [137]:
for k, v in resp.items():
    print(f'=== {k} ===:\n{v}\n')

=== synopsis ===:
"Tragedy at Sunset on the Beach" is a gripping drama that unfolds as a group of friends gather for a relaxing evening on the beach. As the sun sets and tensions rise, long-held secrets and betrayals come to light, leading to a series of heartbreaking events that will change their lives forever.

The play delves into themes of love, friendship, and the consequences of choices made in the heat of the moment. As the protagonists struggle to come to terms with the tragedy that has befallen them, they must navigate their own guilt and grief while grappling with the harsh reality of loss.

"Tragedy at Sunset on the Beach" is a powerful exploration of the fragility of human relationships and the unforeseen events that can shatter them in an instant. It serves as a reminder that even in the most idyllic settings, darkness can descend without warning, forever altering the course of our lives.

=== review ===:
"Tragedy at Sunset on the Beach" is a haunting and emotionally charg

For scenarios involving more complex sequences, the [**SequentialChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.sequential.SequentialChain.html#langchain.chains.sequential.SequentialChain) method comes into play. This allows for multiple inputs and outputs. Consider a case where we need a synopsis based on a play's title and era. Here's how we might set it up:

In [139]:
from langchain_openai import OpenAI
from langchain.chains import LLMChain, SequentialChain
from langchain.prompts import PromptTemplate

llm = OpenAI(temperature=0.7)

synopsis_template = "You are a playwright. Given the title of play and the era it is set in, it is your job to write a synopsis for that title.\n\nTitle: {title}\nEra: {era}\nPlaywright: This is a synopsis for the above play:"
synopsis_prompt_template = PromptTemplate(input_variables=["title", "era"], template=synopsis_template)
synopsis_chain = LLMChain(llm=llm, prompt=synopsis_prompt_template, output_key="synopsis")

review_template = "You are a play critic from the New York Times. Given the synopsis of play, it is your job to write a review for that play.\n\nPlay Synopsis:\n{synopsis}\nReview from a New York Times play critic of the above play:"
prompt_template = PromptTemplate(input_variables=["synopsis"], template=review_template)
review_chain = LLMChain(llm=llm, prompt=prompt_template, output_key="review")

overall_chain = SequentialChain(
    chains=[synopsis_chain, review_chain],
    input_variables=["era", "title"],
    output_variables=["synopsis", "review"],
    verbose=True,
)

In [141]:
resp = overall_chain.invoke({"title": "Tragedy at sunset on the beach", "era": "Victorian England"})



> Entering new SequentialChain chain...

> Finished chain.


In [142]:
resp

{'title': 'Tragedy at sunset on the beach',
 'era': 'Victorian England',
 'synopsis': ' \n\n"Tragedy at Sunset on the Beach" is a gripping tale set in the opulent world of Victorian England. The play follows the aristocratic Lawrence family as they escape the hustle and bustle of London society for a peaceful vacation on the shores of the English coast. As the sun sets on the idyllic beach, tensions rise among the family members, revealing long-held secrets and hidden desires. But when a mysterious figure appears on the beach, the family\'s peaceful getaway turns into a nightmarish tragedy. With betrayal, love, and revenge at its core, this play delves into the dark underbelly of Victorian society and the consequences of keeping up appearances. Will the Lawrence family be able to survive the night, or will the sunset on the beach be their final curtain call? Don\'t miss this thrilling and heartbreaking tragedy that will leave you on the edge of your seat until the very end.',
 'review'

<b>In scenarios where you want to maintain context throughout a chain or for a later part of the chain, [**SimpleMemory**](https://python.langchain.com/api_reference/langchain/memory/langchain.memory.simple.SimpleMemory.html#langchain.memory.simple.SimpleMemory) can be used. This is particularly useful for managing complex input/output relationships</b>. For instance, in a scenario where we want to generate social media posts based on a play's title, era, synopsis, and review, [**SimpleMemory**](https://python.langchain.com/api_reference/langchain/memory/langchain.memory.simple.SimpleMemory.html#langchain.memory.simple.SimpleMemory) can help manage these variables:

In [144]:
from langchain.memory import SimpleMemory
from langchain.chains import SequentialChain

template = '''You are a social media manager for a theater company. Given the title of play, the era it is set in, the date, time and location, the synopsis of the play, and the review of the play,

 it is your job to write a social media post for that play.\n\nHere is some context about the time and location of the play:\nDate and Time: {time}\nLocation: {location}\n\nPlay Synopsis:\n{synopsis}\nReview from a New York Times play critic of the above play:\n{review}\n\nSocial Media Post:'''
prompt_template = PromptTemplate(input_variables=["synopsis", "review", "time", "location"], template=template)
social_chain = LLMChain(llm=llm, prompt=prompt_template, output_key="social_post_text")

overall_chain = SequentialChain(
    memory=SimpleMemory(memories={"time": "December 25th, 8pm PST", "location": "Theater in the Park"}),
    chains=[synopsis_chain, review_chain, social_chain],
    input_variables=["era", "title"],
    output_variables=["social_post_text"],
    verbose=True,
)

In [145]:
resp = overall_chain.invoke({"title": "Tragedy at sunset on the beach", "era": "Victorian England"})



> Entering new SequentialChain chain...

> Finished chain.


In [146]:
resp

{'title': 'Tragedy at sunset on the beach',
 'era': 'Victorian England',
 'time': 'December 25th, 8pm PST',
 'location': 'Theater in the Park',
 'social_post_text': '\n\n🎭 Don\'t miss the gripping murder mystery "Tragedy at sunset on the beach" this December 25th, 8pm PST at the Theater in the Park! Set in Victorian England, this play explores the dark secrets and lies of high society in the picturesque town of Brighton. Join Inspector Hathaway as he unravels the mystery of a young woman\'s tragic death, and be prepared for a shocking conclusion. With themes of love, betrayal, and societal expectations, this play is not one to be missed. See you there! #TragedyAtSunset #TheaterInthePark #VictorianMystery #MurderMystery #Drama #Romance'}

In addition to sequential chains, there are specialized chains for working with documents. Each of these chains serves a different purpose, from combining documents to refining answers based on iterative document analysis, to mapping and reducing document content for summarization or re-ranking based on scored responses. These chains can be recreated with LCEL for additional flexibility and customization:
* [**StuffDocumentsChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.combine_documents.stuff.StuffDocumentsChain.html#langchain.chains.combine_documents.stuff.StuffDocumentsChain): combines a list of documents into a single prompt passed to an LLM.
* [**RefineDocumentsChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.combine_documents.refine.RefineDocumentsChain.html#langchain.chains.combine_documents.refine.RefineDocumentsChain): updates its answer iteratively for each document, suitable for tasks where documents exceed the model's context capacity.
* [**MapReduceDocumentsChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.combine_documents.map_reduce.MapReduceDocumentsChain.html#langchain.chains.combine_documents.map_reduce.MapReduceDocumentsChain): applies a chain to each document individually and then combines the results.
* [**MapRerankDocumentsChain**](https://python.langchain.com/api_reference/langchain/chains/langchain.chains.combine_documents.map_rerank.MapRerankDocumentsChain.html#langchain.chains.combine_documents.map_rerank.MapRerankDocumentsChain): scores each document-based response and selects the highest-scoring one.

<a id='sect_5'></a>
## <b><font color='darkblue'>Module V : Memory</font></b> ([back](#agenda))
<b><font size='3ptx'>In LangChain, memory is a fundamental aspect of conversational interfaces, allowing systems to reference past interactions</font>. This is achieved through storing and querying information, with two primary actions: reading and writing. The memory system interacts with a chain twice during a run, augmenting user inputs and storing the inputs and outputs for future reference</b>.

### <b><font color='darkgreen'>Building Memory into a System</font></b>
1. **Storing Chat Messages**: The LangChain memory module integrates various methods to store chat messages, ranging from in-memory lists to databases. This ensures that all chat interactions are recorded for future reference.
2. **Querying Chat Messages**: Beyond storing chat messages, LangChain employs data structures and algorithms to create a useful view of these messages. Simple memory systems might return recent messages, while more advanced systems could summarize past interactions or focus on entities mentioned in the current interaction.

To demonstrate the use of memory in LangChain, consider the [**ConversationBufferMemory**](https://python.langchain.com/api_reference/langchain/memory/langchain.memory.buffer.ConversationBufferMemory.html) class, a simple memory form that stores chat messages in a buffer. Here's an example:

## <b><font color='darkblue'>Supplement</font></b>
* [Google Gemini Pro Usage via Gemini API and LangChain](https://github.com/sugarforever/LangChain-Tutorials/blob/main/LangChain_Google_Gemini_API.ipynb)
* [Medium - Different Chain Types using LangChain](https://medium.com/@shravankoninti/different-chain-types-using-langchain-89cacae6ad1f)